src/Parallel/Serialize.hpp

```diff
@@ -54,3 +55,20 @@ T deserialize(const void* const data) noexcept {  // NOLINT
   reader | result;
   return result;
 }
+
+/*!
+ * \ingroup ParallelGroup
+ * \brief Deserialize an object using PUP.
+ *
+ * \tparam T the type to deserialize to. The type to deserialize as must be
+ * explicitly specified.
+ */
+template <typename T, typename U>
+void deserialize(const gsl::not_null<U*> result,
+            const void* const data) noexcept {  // NOLINT
+  static_assert(std::is_same_v<T, U>,
```

Comment:
do we need this check?  If we do the message should say deserialization instead of serialization

Reply:
I'm not sure we do, but my instinct was to give it the same safety as the `serialize` function -- @nilsdeppe thoughts?

Reply:
What I don't understand is why you need a second type. Why not just do `gsl::not_null<T*>`?

Reply:
It's the same pattern as `serialize` - it forces the explicit template parameter to avoid type inference, presumably because in the past we were worried about mistakes when passing an unintended object type to the serialization.

Reply:
I'm still confused... In the other case it is done because otherwise we have no idea what time to deserialize to. It literally cannot be inferred (can't infer `T` from the return type). In this case you can infer `T` so I see no reason have to specify it manually. So I don't understand how this is the "same pattern"

Reply:
It is the same pattern as `serialize` (not the previous `deserialize`), which has the snippet:
```
template <typename T, typename U>
std::vector<char> serialize(const U& obj) noexcept {
  static_assert(std::is_same<T, U>::value,
                "Explicit type for serialization differs from deduced type");
```
on line 26 of this file.

Reply:
ah, yea okay, sorry... I'm now trying to figure out what I was worried about. I understand the explanation in the documentation, but I never recall encountering any problems with this. Additionally, the Charm++ code (and all of our `pup` functions) don't (and I think can't?) have a check like this. So my vote is: either correct the message or don't bother deducing the type. Both are fine by me. If we decide to remove the explicit template parameter in the future it would be easy to do anyway since it would only need to be changed in this file. I do generally agree that the type safety guarantees/checks should be the same between serialize and deserialize. However, removing it from `serialize` should also only require a change in this file.